### DISCUSSION TITLE
Getting language composition link

### DISCUSSION THREAD

**Discussion by grrinch** [luyadev/luya#2177] [Questions & Problems] [answered]
Hi,

I'm building a multilingual website and using the built-in composition elements together with the CMS functionality to create language versions.

I came across `LangSwitcher::widget()` but it only generates the language switch widget.

What I am looking for is a method like `Url::to()` but where I can get all alternate links to other language versions (if available).

I looked through the code of `LangSwitcher::widget()` but it doesn't look like there is a method like this - the code there seems to do a lot of stuff in order to get the alternate language links:

Starting with `LangSwitcher::getDataArray()`
```php
foreach (Yii::$app->adminLanguage->getLanguages() as $lang) {
                $array[] = [
                    'lang' => $lang,
                    'item' => Yii::$app->menu->find()->where(['nav_id' => $currentMenuItem->navId])->lang($lang['short_code'])->with('hidden')->one(),
                ];
            }
```
and then processing this in the `LangSwitcher::run()` method:
```php
foreach (self::getDataArray() as $langData) {
            $item = $langData['item'];
            $lang = $langData['lang'];
            $isActive = $currentLang == $lang['short_code'];
            if ($item) {
                if ($item->type == NavItem::TYPE_MODULE && !empty($rule)) {
                    $routeParams = [$rule['route']];
                    foreach ($rule['params'] as $key => $value) {
                        $routeParams[$key] = $this->findUrlRuleParamValue($lang['short_code'], $key, $value);
                    }
                    $compositionObject = Yii::createObject(Composition::class);
                    // https://github.com/luyadev/luya-module-cms/issues/48
                    // $compositionObject->off(Composition::EVENT_AFTER_SET);
                    $compositionObject['langShortCode'] = $lang['short_code'];
                    $link = Yii::$app->urlManager->createMenuItemUrl($routeParams, $item->id, $compositionObject);
                } else {
                    $link = $item->link;
                }
            } else {
                $link = Yii::$app->urlManager->prependBaseUrl($lang['short_code']);
            }

            $items[$lang['short_code']] = $this->generateHtml($this->ensureHostInfo($link, $lang), $isActive, $lang);
            unset($item, $lang);
        }
```

So is this the only way to get all the alternate links or is there a simple composition method for this?

Couldn't find anything in the documentation and it just seems pretty chaotic to run with the above.

Thanks for any help!

**Comment by nadar** [comment, marked as answer]
Hi @grrinch 

Would you like to link to a module route or to a cms page from the menu? 

> for is a method like Url::to()

This looks like you are in view file or a controller? Are URL Rules used? Please make a full example where you like to link at.

**Reply by grrinch** [reply]
It's from the view scope.

Technically, I want to be able to just grab an alternate URL based on a language.

It could be either module or just a CMS page.

Say, something like this:
```php
$url['de'] = Url::to($current_page, ['lang' => 'de']);
```

I just tested that this will work for the CMS page (where you have the id available):
```php
$url['de'] = Yii::$app->menu->find()->where(['nav_id' => $page_id])->lang('de')->with('hidden')->one();
```

But it seems there is no universal way that will work on a module and CMS page independently, based only on the current URL.

Plus, the `LangSwitcher::widget()` generates incorrect links to modules (not their canonical versions).

For example it generates:
`domain.com/en/blog/default/index` or `domain.com/en/contact/default/contact`
instead of just the canonical link to module pages
`domain.com/en/blog/` or `domain.com/en/contact/`

**Reply by nadar** [reply]
I understand your request. But there is no general view to Route "just everywhere". Either you are in CMS context, then there are several options you can use to find other pages as you found out above. 

When working with modules and routes its important to know that you have to declare url rules https://luya.io/guide/frontend/urlrules.html#configure-rules and in order to translate those see https://luya.io/guide/frontend/urlrules.html#multilingual-language-patterns in order to create links for another language you can take a look here:

```php
$compositionObject = Yii::createObject(Composition::class);
// https://github.com/luyadev/luya-module-cms/issues/48
// $compositionObject->off(Composition::EVENT_AFTER_SET);
$compositionObject['langShortCode'] = $lang['short_code'];
$link = Yii::$app->urlManager->createMenuItemUrl($routeParams, $item->id, $compositionObject);
```

But yes, i agree we could quickly build a helper in the Url class :+1: i will add this shortly, should not be a big deal. Thanks for the report.

**Reply by nadar** [reply]
https://github.com/luyadev/luya/issues/2178

**Reply by grrinch** [reply]
OK, great, this one is really helpful:
https://luya.io/guide/frontend/urlrules.html#multilingual-language-patterns

I think it's worth adding it as a reference [here](https://luya.io/guide/app/language.html):

![image](https://user-images.githubusercontent.com/5729796/211013425-a293fe94-1d6f-41e0-a5de-488ecee43e49.png)

I think this is going to fix my issues with URLs incorrectly generated by `LangSwitcher::widget()`.

Thanks!

**Comment by nadar** [comment]
@grrinch i also want to mention that the task of mixin modules and cms routes is very complex. There are also different options on how to add your module. You can add the module by using a module block or declare a page as module, i assume you saw this in the CMS Admin UI. This can also change the behavior of how you generate and interact with routes.

**Reply by grrinch** [reply]
Yes, makes sense. In my case I declare pages as modules in Admin UI.

Think adding composition elements to UrlRule in the module config:

```php
public $urlRules = [
    [
        'pattern' => 'estore/basket',
        'route' => 'estore/basket/default',
        'composition' => [
            'fr' => 'estore/panier',
            'de' => 'estore/warenkorb'
        ]
    ],
];
```

It should sort everything out for me!

**Reply by nadar** [reply]
I have added an example for this when using `Url::toRoute()` (see https://github.com/luyadev/luya.io/commit/9c1d21cc1e16aa28a8805ec55f89ebb28b7b301e)
